pkg/spoke/controllers/finalizercontroller/manifestwork_finalize_controller.go

```diff
@@ -0,0 +1,143 @@
+package finalizercontroller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	workv1client "github.com/open-cluster-management/api/client/work/clientset/versioned/typed/work/v1"
+	workinformer "github.com/open-cluster-management/api/client/work/informers/externalversions/work/v1"
+	worklister "github.com/open-cluster-management/api/client/work/listers/work/v1"
+	"github.com/open-cluster-management/work/pkg/helper"
+	"github.com/open-cluster-management/work/pkg/spoke/controllers"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+)
+
+// ManifestWorkFinalizeController handles cleanup of manifestwork resources before deletion is allowed.
+type ManifestWorkFinalizeController struct {
+	manifestWorkClient        workv1client.ManifestWorkInterface
+	manifestWorkLister        worklister.ManifestWorkNamespaceLister
+	appliedManifestWorkClient workv1client.AppliedManifestWorkInterface
+	appliedManifestWorkLister worklister.AppliedManifestWorkLister
+	hubHash                   string
+	rateLimiter               workqueue.RateLimiter
+}
+
+func NewManifestWorkFinalizeController(
+	recorder events.Recorder,
+	manifestWorkClient workv1client.ManifestWorkInterface,
+	manifestWorkInformer workinformer.ManifestWorkInformer,
+	manifestWorkLister worklister.ManifestWorkNamespaceLister,
+	appliedManifestWorkClient workv1client.AppliedManifestWorkInterface,
+	appliedManifestWorkInformer workinformer.AppliedManifestWorkInformer,
+	hubHash string,
+) factory.Controller {
+
+	controller := &ManifestWorkFinalizeController{
+		manifestWorkClient:        manifestWorkClient,
+		manifestWorkLister:        manifestWorkLister,
+		appliedManifestWorkClient: appliedManifestWorkClient,
+		appliedManifestWorkLister: appliedManifestWorkInformer.Lister(),
+		hubHash:                   hubHash,
+		rateLimiter:               workqueue.NewItemExponentialFailureRateLimiter(5*time.Millisecond, 1000*time.Second),
+	}
+
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
```

Comment:
you need to trigger based on appliedmanifestwork changes too. otherwise you can miss a delete of appliedmanifestwork and forget to create a new instance.

Reply:
nm, its the next line down